v2/batcher.go

```diff
@@ -288,6 +309,88 @@ func (r *Batcher) incTarget(val int) {
 	} // else is val=0, do nothing
 }
 
+func (r *Batcher) tryReserveBatchSlot() bool {
+	if r.maxConcurrentBatches == 0 {
+		return true
+	}
+	select {
+	case r.inflight <- struct{}{}:
+		return true
+	default:
+		return false
+	}
+}
+
+func (r *Batcher) releaseBatchSlot() {
+	if r.maxConcurrentBatches > 0 {
+		<-r.inflight
+	}
+}
+
+func (r *Batcher) confirmInflightIsZero() bool {
+	isZero := true
+	for {
+		select {
+		case <-r.inflight:
+			isZero = false
+		default:
+			return isZero
+		}
+	}
+}
+
+func (r *Batcher) Inflight() uint32 {
+	return uint32(len(r.inflight))
+}
+
+func (r *Batcher) processBatch(watcher IWatcher, batch []IOperation) {
+	if len(batch) == 0 {
+		return
+	}
+	r.lastFlushWithRecords = time.Now()
+
+	// raise event
+	if r.emitBatch {
+		r.emit(BatchEvent, len(batch), "", batch)
+	}
+
+	go func() {
+
+		// increment an attempt
+		for _, op := range batch {
+			op.MakeAttempt()
+		}
+
+		// process the batch
+		waitForDone := make(chan struct{})
+		go func() {
+			defer close(waitForDone)
+			watcher.ProcessBatch(batch)
+		}()
+
+		// the batch is "done" when the ProcessBatch func() finishes or the maxOperationTime is exceeded
+		maxOperationTime := r.maxOperationTime
+		if watcher.MaxOperationTime() > 0 {
+			maxOperationTime = watcher.MaxOperationTime()
+		}
+		select {
+		case <-waitForDone:
+		case <-time.After(maxOperationTime):
+		}
+
+		// decrement target
+		var total int = 0
+		for _, op := range batch {
+			total += int(op.Cost())
+		}
+		r.incTarget(-total)
+
+		// remove from inflight
+		r.releaseBatchSlot()
```

Comment:
Any cases where this might not be reached?

Reply:
The select blocks until MaxOperationTime (Watcher if it has one, Batcher if not; 1 min default) or the goroutine wrapping the callback exits (even if it were a panic). So the only way this block doesn't get hit is if the programmer has chosen an effectively forever MaxOperationTime (like 100 years) and have blocked in the callback. Even if that happens though, at the AuditInterval those will be cleaned up provided the Batcher MaxOperationTime is exceeded. So again, they could defeat this by choosing a 100 year Batcher MaxOperationTime.

SUMMARY: I feel like it is as protected as it can be outside of the dev choosing to defeat it.